gcp_variant_transforms/libs/generate_report.py

```diff
@@ -0,0 +1,127 @@
+
+  _HEADER_LINE = ['ID', 'Conflicts', 'Proposed Solution']
+  _REPORT_NAME = 'header_conflicts_report.csv'
+  _NO_CONFLICTS_MESSAGE = ['No conflicts found.']
+  _NO_SOLUTION_MESSAGE = 'Not resolved'
+
+  def __init__(self, directory=''):
+    # type: (str) -> None
+    """Initializes ``ConflictsReporter`` object.
+
+    Args:
+      directory: The location where the conflicts report is saved.
+    """
+    self._directory = directory
+
+  def report_conflicts(self, header_definitions, resolved_header=None):
+    # type: (VcfHeaderDefinitions, VcfHeader) -> None
+    """Generates a report.
+
+    Combines the conflicts extracted from ``header_definitions`` and their
+    resolutions from ``resolved_header`` to generate a conflicts report.
+    """
+    resolved_header = resolved_header or VcfHeader()
+    format_conflicts = self._extract_conflicts(header_definitions.formats)
+    info_conflicts = self._extract_conflicts(header_definitions.infos)
+    format_headers = resolved_header.formats
+    info_headers = resolved_header.infos
+    contents = list()
+    contents.extend(self._generate_contents(format_conflicts, format_headers))
+    contents.extend(self._generate_contents(info_conflicts, info_headers))
+    self._write_to_report(contents)
+
+  def _extract_conflicts(
+      self,
+      definitions  # type: Dict[str, Dict[Definition, List[str]]]
+      ):
+    # type: (...) -> Dict[str, Dict[Definition, List[str]]]
+    """Extracts the fields that have conflicted definitions."""
+    # len(v) > 1 means there are conflicted definitions for this field.
+    return dict([(k, v) for k, v in definitions.items() if len(v) > 1])
+
+  def _generate_contents(
+      self,
+      conflicts,  # type: Dict[str, Dict[Definition, List[str]]]
+      headers  # type: Dict[str, Dict[str, Union[str, int]]
+      ):
+    # type: (...) -> List[str]
+    """Generates the report contents.
+
+    The conflicted definitions, the file names and the resolutions are included
+    in the contents.
+    """
+    contents = list()
+    for field_id, definitions_to_files_map in conflicts.iteritems():
+      row = [
+          field_id,
+          self._extract_definitions_and_file_names(definitions_to_files_map),
+          self._extract_resolution(headers, field_id)
+      ]
+      contents.append(row)
+    return contents
+
+  def _extract_definitions_and_file_names(self, definition_to_files_map):
+    # type: (Dict[Definition, List[str]]) -> str
+    conflict_definitions = []
+    for definition, file_names in definition_to_files_map.iteritems():
+      definition = self._format_definition(definition.num, definition.type)
+      conflict_definitions.append(definition + ' in ' + str(file_names))
+    return ' '.join(conflict_definitions)
+
+  def _extract_resolution(self, header, filed_id):
+    # type: (Dict[str, Dict[str, Union[str, int]]], str) -> str
+    if filed_id not in header:
+      return self._NO_SOLUTION_MESSAGE
```

Comment:
Please comment which scenarios this can happen (I suspect we'll always have a resolution unless no `resolved_header` is provided?).